pkg/trace/api: add telemetry proxy endpoint

pawelchcki opened the pull request:
### What does this PR do?
This PR adds new endpoint into trace-agent to allow easy proxying of telemetry requests into our intake.
Its heavily inspired by the Profiling proxy, ~trying to reuse some code where possible - and make it compatible with most of other proxy implementations.~

~Its possible with minor changes (+10 lines) it can be adapted to work for both debugger and profiling proxy. Making other implementations obsolete~

Its also a multi endpoint proxy, meaning it will broadcast the requests into all configured "additional endpoints" if additional endpoints are configured


### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

Write here in detail how you have tested your changes
and instructions on how this should be tested in QA.

Describe or link instructions to set up environment 
for testing, if the process requires more than just
running the agent on one of the supported platforms.

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

gbbr commented on pkg/config/apm.go:
```diff
@@ -11,6 +11,12 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+func setupAPMTelemetry(config Config) {
```
I think all of this can go with the rest (just like the profiler and debugger). It doesn't warrant a separate function.

gbbr commented on pkg/trace/api/proxy.go:
```diff
@@ -0,0 +1,192 @@
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type proxyEndpointsConfig struct {
```
Why not `telemetryProxy`?

pawelchcki replied:
This code was meant to try to be plugin replacement for other proxies in the agent right now. 
The other proxies are not that much different, its mostly superficial and I think could be unified with right abstraction:
The differences mainly are;
- URL building, and handling, some forward `prefix/` some forward full Path.
- Forwarding, sending data to a single DC or multiple via additional_endpoints - proxyEndpointsConfig if additional endpoints cfg option are not specified, it will only ever send to a single DC
- Attaching metadata, either they do not attach metadata or they do, in 2-3 different ways - this can be handled via additionalDirector thats executed on every request

In short making this proxy  work for  ProfilingProxy only requires appropriate data in proxyEndpointsConfig, and a custom additionalDirector setting correct metadata. Otherwise the code will do the same things to the request.

AppSec proxy, requires additionalDirector, and a change to disabledMessage.

Same for debugger proxy

I haven't created a PR to unify this, but at least I was hoping this change will help other teams in the future. So the code reflects that to an extent.

gbbr commented on pkg/trace/api/proxy.go:
```diff
@@ -0,0 +1,192 @@
+// is proxied back to the client, while for all aditional endpoints the
+// response is discarded. There is no de-duplication done between endpoint
+// hosts or api keys.
+type proxyMultiTransport struct {
```
I think I preferred the previous name much more. I don't see the point of adding the prefix `proxy*`.

```suggestion
type multiTransport struct {
```

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,11 @@
+package api
+
+var telemetryProxyEndpointsConfig = proxyEndpointsConfig{
+	name:                      "telemetry",
+	enabledConfig:             "apm_config.telemetry.enabled",
+	ddURLConfig:               "apm_config.telemetry.dd_url",
+	additionalEndpointsConfig: "apm_config.telemetry.additional_endpoints",
```
```suggestion
	name:                      "telemetry",
```
What if we keep just `telemetry`? The rest can be deduced from it using `apm_config.<name>.{enabled,dd_url,additional_endpoints}`

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,11 @@
+	ddURLConfig:               "apm_config.telemetry.dd_url",
+	additionalEndpointsConfig: "apm_config.telemetry.additional_endpoints",
+	urlTemplate:               "https://instrumentation-telemetry-intake.%s",
+	defaultURL:                "https://instrumentation-telemetry-intake.datadoghq.com/",
```
I think we just need the `URLTemplate` and a `DefaultSite` (`datadoghq.com`)

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,11 @@
+package api
+
+var telemetryProxyEndpointsConfig = proxyEndpointsConfig{
+	name:                      "telemetry",
```
Can we capitalise all config fields? They look odd in private.

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,11 @@
+	additionalEndpointsConfig: "apm_config.telemetry.additional_endpoints",
+	urlTemplate:               "https://instrumentation-telemetry-intake.%s",
+	defaultURL:                "https://instrumentation-telemetry-intake.datadoghq.com/",
+	additionalDirector:        nil,
```
You can remove this field.

gbbr commented on pkg/trace/api/proxy.go:
```diff
@@ -0,0 +1,192 @@
+	additionalDirector        func(req *http.Request)
+}
+
+func (pec *proxyEndpointsConfig) proxyRoundTripper(client http.RoundTripper, apiKey string) (http.RoundTripper, error) {
```
```suggestion
func (pec *proxyEndpointsConfig) roundTripper(client http.RoundTripper, apiKey string) (http.RoundTripper, error) {
```

gbbr commented on pkg/trace/api/proxy.go:
```diff
@@ -0,0 +1,192 @@
+	}, err
+
+}
+func (pec *proxyEndpointsConfig) proxyEndpoints(apiKey string) (urls []*url.URL, apiKeys []string, err error) {
```
```suggestion
func (pec *proxyEndpointsConfig) endpoints(apiKey string) (urls []*url.URL, apiKeys []string, err error) {
```

gbbr commented on pkg/trace/api/proxy.go:
```diff
@@ -0,0 +1,192 @@
+	return pec.newForwardingProxy(transport, r.conf.Hostname, r.conf.DefaultEnv)
+}
+
+func (pec *proxyEndpointsConfig) proxyDisabledHandler(err error) http.Handler {
```
```suggestion
func (pec *proxyEndpointsConfig) disabledHandler(err error) http.Handler {
```

gbbr commented on pkg/trace/api/proxy.go:
```diff
@@ -0,0 +1,192 @@
+		return pec.proxyDisabledHandler(err)
+	}
+
+	return pec.newForwardingProxy(transport, r.conf.Hostname, r.conf.DefaultEnv)
```
This should be a function, not a method. The proxy config `pec` should be an argument not a receiver. The return values with accessing private functions like `pec.proxyDisabledHandler` and passing them errors from an outer scope is an odd pattern.

I think this could be more idiomatic and simplified if we got rid of this entire function (`(*HTTPReceiver).forwardingProxyHandler`) and instead just return one call in `endpoints.go`:
```go
return newForwardingProxy(pec, r.conf.Hostname, r.conf.DefaultEnv)
```
This function will then return an `http.Handler` which could be the proxy or the "disabled handler".

gbbr commented on pkg/trace/api/endpoints.go:
```diff
@@ -76,6 +76,12 @@ var endpoints = []endpoint{
+	{
+		Pattern: "/telemetry/proxy/",
+		Handler: func(r *HTTPReceiver) http.Handler {
+			return http.StripPrefix("/telemetry/proxy", r.forwardingProxyHandler(&telemetryProxyEndpointsConfig))
```
```go
return newForwardingProxy(&proxyConfig{
	Name:        "telemetry",
    Hostname:    r.conf.Hostname,
    DefaultEnv:  r.conf.DefaultEnv,
    URLTemplate: "https://instrumentation-telemetry-intake.%s",
    DefaultSite: "datadoghq.com",
})
```
I think this makes a lot more sense. The `telemetry.go` file doesn't help. It just misplaces something that should be readily available to read here.

gbbr reviewed (commented):
For some reason it looks like you tried to create a generic proxy structure which is not re-used by either profiler, nor live debugger (nor appsec).

pawelchcki commented:
@gbbr Hey, sorry for such long response to you review. I've refactored to include most of the suggestsions, and simplify the existing implementation. 
+ added some more tests for multi proxy capability.

if you have some time please take another look 🙇

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,207 @@
+//
+// can be extended in the future to allow supporting more product endpoints
+// by simply parametrizing metric tags, and logger names
+type telemetryMultiTransport struct {
```
Can we rename this to `telemetryProxy` and have it contain everything including the above config (dropping `proxyEndpointsConfig` and the rest)? That config above is just a remnant of the previous attempt to de-dupe things, which we've now given up.

pawelchcki replied:
I still like that remnant, as it allows me to configure everything in one self contained place. 

I was thinking it will get this feedback though, Let me try to find an alternative.

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,207 @@
+// by simply parametrizing metric tags, and logger names
+type telemetryMultiTransport struct {
+	rt                http.RoundTripper
+	baseTarget        target
+	additionalTargets []target
```
I don't think we need a new `target` type. Can we just use the same approach as the other `multiTransport` here: https://github.com/DataDog/datadog-agent/blob/eb6a11858279d402982d93bfc00bfeb32072b3cc/pkg/trace/api/profiles.go#L133-L137 ?

Also, is there an opportunity here to re-use the other `multiTransport`?

pawelchcki replied:
I'd have to refactor it a bit. Because the list of target ursl and list of target keys, is IMO inferior to a list of target structs  and an explicit main struct.

pawelchcki commented:
@gbbr I think the code is ready for another review round. 

I went ahead I moved most stuff to a separate package mainly for three reasons:
1) there's already a fair bit of code I've been trying to segment in the single file, splitting it to 3 files (config, transport and integration) made sense to me 
2) I don't like adding artificial prefixes to functions structs if a language has namespacing feature :D (too much trauma from C programming ;) ) 
3) We'll most probably add some telemetry reporting to the trace agent itself - at some point. pkg/trace/telemetry seems like a good place for that code to live in  

+ we might add some backoff+limiting feature to main transport. We need more testing and data first to know if thats a good idea.

pawelchcki committed:
Add agent telemetry proxy endpoint

gbbr commented on pkg/trace/telemetry/config_test.go:
```diff
@@ -0,0 +1,85 @@
+		}
+	})
+
+	t.Run("default site with malformed additional endpoints", func(t *testing.T) {
```
Just one word for test names please: https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#tests

gbbr commented on pkg/trace/telemetry/config.go:
```diff
@@ -0,0 +1,63 @@
+package telemetry
```
This does not warrant a separate package (just like we don't have separate packages for the other proxies either: profiler, live debugger, etc). It's not a stand alone unit that is used anywhere else. Can we please put everything inside `pkg/trace/api/telemetry.go`?

pawelchcki committed:
Move telemetry config into trace/config

pawelchcki committed:
Make telemetry use AgentConfig

gbbr commented on pkg/trace/config/config.go:
```diff
@@ -193,6 +196,8 @@ func New() *AgentConfig {
 
 		DDAgentBin:   defaultDDAgentBin,
 		OTLPReceiver: &OTLP{},
+		TelemetryConfig: &TelemetryConfig{
+			Endpoints: []*Endpoint{{Host: "https://instrumentation-telemetry-intake.datadoghq.com"}}},
```
```suggestion
		TelemetryConfig: &TelemetryConfig{
			Endpoints: []*Endpoint{{Host: "https://instrumentation-telemetry-intake.datadoghq.com"}},
		},
```

gbbr commented on releasenotes/notes/add-trace-agent-telemetry-endpoint-f398bcd47e637081.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features:
+  - |
+    APM: Add telemetry proxy endpoint to Trace Agent to allow Instrumentation Libraries to submit telemetry data
```
```suggestion
    APM: Add proxy endpoint to allow Instrumentation Libraries to submit telemetry data.
```
The `APM: ` prefix already makes it clear that it's the trace-agent.

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -358,6 +368,62 @@ func (c *AgentConfig) applyDatadogConfig() error {
+			extra := config.Datadog.GetStringMapStringSlice(additionalEndpoitnsCfg)
+			for endpoint, keys := range extra {
+				u, err := url.Parse(endpoint)
+
```
If we're going to use `additional_endpoints`, the logic should be the same as elsewhere. Can we simply duplicate it or create a function which does it? Maintaining backwards compatibility. Check https://github.com/DataDog/datadog-agent/blob/2373feabd39cf4148fa0fcc7fb2235bebb2dc220/pkg/trace/config/apply.go#L229-L240

pawelchcki replied:
Ok, I changed plans. Seeing how Endpoint.Host is actually used more as an URL in trace/writer 

I'm now tempted to refactor Endpoint.Host into Endpoint.URL. but thats a story for a separate PR - I think :)

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -369,6 +435,7 @@ func (c *AgentConfig) applyDatadogConfig() error {
 			}
 		}
 	}
+
```
Please remove it. Check point 3 here https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#general-guidelines

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -358,6 +368,62 @@ func (c *AgentConfig) applyDatadogConfig() error {
+			}
+		}
+	}
+
+	// Obfuscation config
```
Same here. There is no need for the comment, nor the spacing.

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -358,6 +368,62 @@ func (c *AgentConfig) applyDatadogConfig() error {
+	c.TelemetryConfig = &TelemetryConfig{
+		Enabled: config.Datadog.GetBool("apm_config.telemetry.enabled"),
+	}
+	if c.TelemetryConfig.Enabled {
+		site := config.Datadog.GetString("site")
+		if site == "" {
+			site = defaultSite
+		}
+
+		main := telemetryEndpointPrefix + site
+		if v := config.Datadog.GetString("apm_config.telemetry.dd_url"); v != "" {
+			main = v
+		}
+
+		u, err := url.Parse(main)
+		if err != nil {
+			// if the main intake URL is invalid we don't use additional endpoints
+			log.Errorf("error parsing main telemetry intake URL %s: %v", main, err)
+			c.TelemetryConfig.Enabled = false
+		} else {
+			c.TelemetryConfig.Endpoints = []*Endpoint{{
+				APIKey: c.Endpoints[0].APIKey,
+				Host:   u.Host,
+			}}
+		}
```
```suggestion
	if c.TelemetryConfig.Enabled {
		site := config.Datadog.GetString("site")
		if site == "" {
			site = defaultSite
		}
		main := telemetryEndpointPrefix + site
		if v := config.Datadog.GetString("apm_config.telemetry.dd_url"); v != "" {
			main = v
		}
		u, err := url.Parse(main)
		if err != nil {
			// if the main intake URL is invalid we don't use additional endpoints
			log.Errorf("error parsing main telemetry intake URL %s: %v", main, err)
			c.TelemetryConfig.Enabled = false
		} else {
			c.TelemetryConfig.Endpoints = []*Endpoint{{
				APIKey: c.Endpoints[0].APIKey,
				Host:   u.Host,
			}}
		}
```
Sorry to be nitty, just trying to keep consistent.

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -358,6 +368,62 @@ func (c *AgentConfig) applyDatadogConfig() error {
 		MaxRequestBytes: c.MaxRequestBytes,
 	}
 
+	// Instrumentation Telemetry config
```
Same thing here. Comment is not needed.

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -358,6 +368,62 @@ func (c *AgentConfig) applyDatadogConfig() error {
+		Enabled: config.Datadog.GetBool("apm_config.telemetry.enabled"),
+	}
+	if c.TelemetryConfig.Enabled {
+		site := config.Datadog.GetString("site")
```
I'm 100% sure there has to already be a helper method for this. If not, let's make one. https://github.com/DataDog/datadog-agent/blob/2373feabd39cf4148fa0fcc7fb2235bebb2dc220/pkg/trace/config/apply.go#L219-L227

pawelchcki replied:
There was config.GetMainEndpoint function I was able to use

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -28,6 +28,10 @@ import (
 
+// telemetryEndpointPrefix is the URL prefix for Instrumentation telemetry endpoint
+const telemetryEndpointPrefix = "https://instrumentation-telemetry-intake."
+const defaultSite = "datadoghq.com"
```
Delete this please and use `(pkg/config).DefaultSite`

pawelchcki replied:
👍 Was looking for previously defined default site, and couldn't find it. 🙈 - On it

pawelchcki replied:
Done

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -28,6 +28,10 @@ import (
 // apiEndpointPrefix is the URL prefix prepended to the default site value from YamlAgentConfig.
 const apiEndpointPrefix = "https://trace.agent."
```
We can group these:
```suggestion
const (
    apiEndpointPrefix = "https://trace.agent."
    telemetryEndpointPrefix = "https://instrumentation-telemetry-intake."
)
    
```

gbbr commented on pkg/trace/api/telemetry_test.go:
```diff
@@ -0,0 +1,158 @@
+	}
+}
+
+func TestTelemetryProxyMultipleEndpoitns(t *testing.T) {
```
```suggestion
func TestTelemetryProxyMultipleEndpoints(t *testing.T) {
```
Please check for this typo. You've made it in other places too 🙂

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,135 @@
+			http.Error(w, msg, http.StatusMethodNotAllowed)
+		})
+	}
+
```
```suggestion
```
Please remove this line.

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,135 @@
+// If any request fails, the error will be logged. Only main target's
+// error will be propagated via return value
+func (m *telemetryMultiTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if len(m.Endpoints) == 1 {
+		return m.roundTrip(req, m.Endpoints[0])
+	}
+
+	slurp, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	newreq := req.Clone(req.Context())
+	newreq.Body = ioutil.NopCloser(bytes.NewReader(slurp))
+	rresp, rerr := m.roundTrip(newreq, m.Endpoints[0])
+
+	for _, endpoint := range m.Endpoints[1:] {
```
```suggestion
	if len(m.Endpoints) == 1 {
		return m.roundTrip(req, m.Endpoints[0])
	}
	slurp, err := ioutil.ReadAll(req.Body)
	if err != nil {
		return nil, err
	}
	newreq := req.Clone(req.Context())
	newreq.Body = ioutil.NopCloser(bytes.NewReader(slurp))
	rresp, rerr := m.roundTrip(newreq, m.Endpoints[0])
	for _, endpoint := range m.Endpoints[1:] {
```

gbbr reviewed (commented):
This is really looking great. I love how much we've managed to simplify. I have a lot of nits. Hoping it won't be too annoying. Please bear with me.

pawelchcki committed:
Update pkg/trace/config/config.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update releasenotes/notes/add-trace-agent-telemetry-endpoint-f398bcd47e637081.yaml

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/config/apply.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/api/telemetry_test.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/config/apply.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
fix typo Endpoints

pawelchcki committed:
merge both additional_endpoints implementations

pawelchcki committed:
rempove unnecessary spaces

pawelchcki committed:
use config.GetMainEndpoint to configure main endpoint

pawelchcki committed:
Reuse common code in config apply

pawelchcki commented on pkg/trace/config/config_test.go:
```diff
@@ -33,6 +33,7 @@ func TestConfigHostname(t *testing.T) {
 	t.Run("fail", func(t *testing.T) {
 		defer cleanConfig()()
 		config.Datadog.Set("apm_config.dd_agent_bin", "/not/exist")
+		config.Datadog.Set("cmd_port", "-1")
```
added this to make the test pass on my system (I have dd agent running in the background). 
I think its a worthy addition, to make sure the test will not actually try to connect to the agent, If we only want to test the fallback function

pawelchcki committed:
Remove unnecessary change of logging severity

pawelchcki commented:
Thanks for the review! 
I've pushed the requested changes, 
I think it looks good. Possible improvement could be to refactor Endpoint struct to use URLs instead of Host string (which holds a URL as a string, when used in Trace/writer and only host when used in telemetry. 

I haven't yet attempted it, but I expect it to be big change especially in trace/writer. So probably best to be done in a separate PR.

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -357,7 +365,39 @@ func (c *AgentConfig) applyDatadogConfig() error {
 		GRPCPort:        grpcPort,
 		MaxRequestBytes: c.MaxRequestBytes,
 	}
+	if c.TelemetryConfig == nil {
```
You could make sure it's never `nil` by initialising it in `New()`, along with its defaults (such as the endpoints below).

pawelchcki replied:
Its initialized in New. But I've seen examples in the apply that already don't seem to trust anything being properly initialized, so I thought I'll follow suite. 

Also it seems apply mostly replaces everything either way, so New() seems kind of unnecessary :) 

I'll remove it though, as it as you point out is not strictly necessary addition

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -357,7 +365,39 @@ func (c *AgentConfig) applyDatadogConfig() error {
+			APIKey: c.Endpoints[0].APIKey,
+		}}
+		endpoints = appendAdditionalEndpoints(endpoints, "apm_config.telemetry.additional_endpoints")
+		// convert any URLs into Hostnames
```
Shouldn't we use URLs here too? To be consistent. I'd expect all `additional_endpoints` functionality to work the same everywhere. It's used in many places in the agent (even outside APM). Doing it differently here could create confusion. Am I missing something?

pawelchcki replied:
IMO There are two problems existing with Endpoint.Host representation. 
1) Its named Host, but holds a URL :) 
2) it holds a string, that then needs to be parsed into URL representation. If it is malformed, it will cause an error to show up later in the stack. 

I think any URL representation errors should be surfaced at the config parsing time, the same way as we check if APIKeys exist for defined additional endpoints.

pawelchcki replied:
I can change it to only validate its a valid url, and not change the value here.
Though I'll have to parse it another place in the code again.

pawelchcki replied:
Done, I've moved the conversion to the telemetry initialization. But kept the validation. @gbbr  Let me know what you think

pawelchcki replied:
Ok per discussion on slack. I removed any validation from applyConfig. And will only rely on validation in telemetry proxy component. 
I've added a few new tests there to cover this behaviour, and cleaned up the test setup a little to make it easier on the eyes (at least mine :) )

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -466,7 +506,7 @@ func (c *AgentConfig) applyDatadogConfig() error {
 		profilingSite := config.Datadog.GetString("internal_profiling.profile_dd_url")
 		if profilingSite == "" {
-			profilingSite = site
+			profilingSite = config.Datadog.GetString("site")
```
This change seems out of scope, or for another PR. I don't think we have a bug here, or do we?

pawelchcki replied:
This is because of the refactoring. The site variable was defined earlier. But it got replaced with config.GetMainEndpoint.

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/config/apply.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/config/apply.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki commented on pkg/trace/config/apply.go:
```diff
@@ -199,9 +208,28 @@ type WriterConfig struct {
 	if len(c.Endpoints) == 0 {
-		c.Endpoints = []*Endpoint{{}}
+		c.Endpoints = []*Endpoint{{
+			Host: apiEndpointPrefix + config.DefaultSite,
```
Yes, but 
I noticed previous code (from main branch): 
```
if len(c.Endpoints) == 0  {
               c.Endpoints = []*Endpoint{{}}
}
```
Would end up producing invalid configuration, with Only API key set and without any hostname.

Not knowing all of the possible configurations the applyDatadogConfig can be called in, I thought its safer to fix this initialization to produce a valid config.

But if I understand you correctly the original check is not necessary. So I think we should keep the change as is. Or remove unnecessary original check.

pawelchcki replied:
Or leave it as is :) 
I've remove the change, since it doesn't impact anything at the moment anyway :)

pawelchcki committed:
Update pkg/trace/config/apply.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Remove unnecessary config resets in applyDatadogConfig

pawelchcki committed:
Only verify endpoint in apply config. use url.Parse to extract target host at telemetry initialization.

pawelchcki committed:
Use defer to time the roundtrip, change the metric name to be distinct

pawelchcki committed:
Allow telemetry config apply to use malformed urls.

Verify URLs in telemetry proxy component

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,169 @@
+	return rresp, rerr
+}
+
+func telemetryEndpoints(conf config.AgentConfig) (endpoints []*config.Endpoint) {
```
Is this used anywhere?

pawelchcki replied:
Uh sorry - this shouldn't have been commited.

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	// extract and validate Hostnames from configured endpoints
+	endpoints := []*config.Endpoint{}
```
```suggestion
	var endpoints []*config.Endpoint
```

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,169 @@
+	for _, endpoint := range r.conf.TelemetryConfig.Endpoints {
+		u, err := url.Parse(endpoint.Host)
+		if err != nil {
+			log.Errorf("Error parsing Telemetry endpoint: %s", endpoint.Host)
```
```suggestion
			log.Errorf("Error parsing apm_config.telemetry endpoint: %s", endpoint.Host)
```

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -358,6 +366,16 @@ func (c *AgentConfig) applyDatadogConfig() error {
+			APIKey: c.Endpoints[0].APIKey,
+		}}
+		c.TelemetryConfig.Endpoints = appendEndpoints(c.TelemetryConfig.Endpoints, "apm_config.telemetry.additional_endpoints")
+		if len(c.TelemetryConfig.Endpoints) > 0 {
+			c.TelemetryConfig.Enabled = true
+		}
```
You can remove this now

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
remove unused code

pawelchcki committed:
Fix tests failing when API key is not set

gbbr commented on pkg/trace/api/telemetry_test.go:
```diff
@@ -0,0 +1,204 @@
+	})() // reset config after the test
+
+	req, rec := newRequestRecorder(t)
+	recv := NewHTTPReceiver(loadConfig(t), nil, nil, nil)
```
Can you use `newTestReceiverFromConfig`?

gbbr commented on pkg/trace/api/telemetry_test.go:
```diff
@@ -0,0 +1,204 @@
+		return nil
+	})
+
+	additionalEndpoints := make(map[string]string)
```
Do you mind removing this variable and defining the map inline below inside `mockConfigMap`? It's easier to read and follow. Plus, the variable is not really used anywhere else.

pawelchcki replied:
Yeah, it turned out I assumed wrong about what are the properties of Go maps. Your suggested way is much better 🙇

gbbr commented on pkg/trace/api/telemetry_test.go:
```diff
@@ -0,0 +1,204 @@
+	// because we use number 2,3 both endpoints must be called to produce 5
+	// just counting number of requests could give false results if first endpoint
+	// was called twice
+	if atomic.LoadUint64(&endpointCalled) != 5 {
```
![Tux, the Linux mascot](https://c.tenor.com/dp_hQBGT0rIAAAAC/think-smart.gif)

pawelchcki replied:
Maths FTW :)

gbbr commented on pkg/trace/config/apply_test.go:
```diff
@@ -59,3 +60,119 @@ func TestSplitTag(t *testing.T) {
+
+	t.Run("keep-malformed", func(t *testing.T) {
+		defer cleanConfig()
+		additionalEndpoints := make(map[string]string)
+		additionalEndpoints["11://test_backend_2.example.com///"] = "test_apikey_2"
+		additionalEndpoints["http://test_backend_3.example.com/"] = "test_apikey_3"
```
```suggestion
		additionalEndpoints := map[string]string{
			"11://test_backend_2.example.com///": "test_apikey_2",
			"http://test_backend_3.example.com/": "test_apikey_3",
		}
```

gbbr commented on pkg/trace/config/apply_test.go:
```diff
@@ -59,3 +60,119 @@ func TestSplitTag(t *testing.T) {
+
+	t.Run("additional-hosts", func(t *testing.T) {
+		defer cleanConfig()
+		additionalEndpoints := make(map[string]string)
+		additionalEndpoints["test_backend_2.example.com"] = "test_apikey_2"
+		additionalEndpoints["test_backend_3.example.com"] = "test_apikey_3"
```
Can you please define the map inline (as suggested in previous comment) in all the tests?

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -216,28 +244,8 @@ func (c *AgentConfig) applyDatadogConfig() error {
 		c.StatsdPort = config.Datadog.GetInt("dogstatsd_port")
 	}
 
-	site := config.Datadog.GetString("site")
```
Very nice cleanup here 🙏🏻

pawelchcki committed:
Update pkg/trace/config/apply_test.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,153 @@
+	if err != nil {
+		return nil, err
+	}
+	newreq := req.Clone(req.Context())
+	newreq.Body = ioutil.NopCloser(bytes.NewReader(slurp))
+	rresp, rerr := m.roundTrip(newreq, m.Endpoints[0])
```
First endpoint is handled in special Way - compared to other ones. 
We'll get the response and error and return it as is. Where as other Endpoints errors are only logged while the response is discarded.

gbbr replied:
```suggestion
	newreq := req.Clone(req.Context())
	newreq.Body = ioutil.NopCloser(bytes.NewReader(slurp))
	// despite the number of endpoints, we always return the response of the first
	rresp, rerr := m.roundTrip(newreq, m.Endpoints[0])
```
In that case, maybe it's worth a comment?

pawelchcki committed:
use inline maps

pawelchcki commented on pkg/trace/api/info_test.go:
```diff
@@ -172,6 +183,7 @@ func TestInfoHandler(t *testing.T) {
 		"/v0.5/traces",
 		"/v0.6/traces",
 		"/profiling/v1/input",
+		"/telemetry/proxy/",
```
I haven't noticed IsEnabled before. It seems like a great addition to endpoints struct! also +1 for consistency

gbbr replied:
Nice. Thanks!

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Use go maps correctly and efficiently

pawelchcki committed:
document helpers and inline loadConfig helper

pawelchcki committed:
remove unnecessary Forwarding proxy function

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Use isEnabled endpoint helper. Simplify telemetrymultitransport implementation

pawelchcki commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,153 @@
+
+		endpoints = append(endpoints, endpoint)
+	}
+	if len(endpoints) == 0 {
```
Since I removed previous handler - now that IsEnabled is used, I think a special handler here would be useful.
I can make the message more generic though.

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,136 @@
+	}
+	if len(endpoints) == 0 {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			log.Errorf("Telemetry proxy couldn't forward a request - no valid endpoints found")
```
This log needs to be outside of the handler. Otherwise it'll show every time someone accesses this. We want to see the error immediately when the handler is initialized.

pawelchcki committed:
rearange methods in telemetry.go

pawelchcki committed:
Merge remote-tracking branch 'origin/main' into pawel/add_simple_telemetry_proxy_endpoint

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,136 @@
+	for _, endpoint := range r.conf.TelemetryConfig.Endpoints {
+		u, err := url.Parse(endpoint.Host)
+		if err != nil {
+			log.Errorf("Error parsing apm_config.telemetry endpoint: %s", endpoint.Host)
```
```suggestion
			log.Errorf("Error parsing apm_config.telemetry endpoint %q: %v", endpoint.Host, err)
```
Maybe the error ends up being useful?

pawelchcki committed:
Incoroprate Gabriel's change

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

gbbr commented on pkg/trace/api/endpoints.go:
```diff
@@ -88,6 +89,13 @@ var endpoints = []endpoint{
 	},
+	{
+		Pattern: "/telemetry/proxy/",
+		Handler: func(r *HTTPReceiver) http.Handler {
```
Can you please make this function all on one line? All of the others are too.

pawelchcki replied:
Its automatically expanded for me to 3 lines because of its length. Are there any line limits ?

gbbr replied:
It's fine then. Leave it as it is.

gbbr commented on pkg/trace/api/telemetry.go:
```diff
@@ -0,0 +1,136 @@
+
+	resp, err := m.Transport.RoundTrip(req)
+	if err != nil {
+		metrics.Count("datadog.trace_agent.telemetry_proxy.roundtrip.errors", 1, tags, 1)
```
```suggestion
		metrics.Count("datadog.trace_agent.telemetry_proxy.error", 1, tags, 1)
```
Or `roundtrip_error`. We should minimise nesting.

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -466,7 +482,7 @@ func (c *AgentConfig) applyDatadogConfig() error {
 		profilingSite := config.Datadog.GetString("internal_profiling.profile_dd_url")
 		if profilingSite == "" {
-			profilingSite = site
+			profilingSite = config.Datadog.GetString("site")
```
This is all wrong. Not because of your PR, because of the documentation of this function. It should work like this, I think: https://github.com/DataDog/datadog-agent/blob/c09f2bc6d3a13e213bcfa4270c7c09fe5a7cceec/pkg/process/config/yaml_config.go#L246-L253

pawelchcki committed:
Update pkg/trace/api/telemetry.go

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

pawelchcki committed:
Properly set profiling site

pawelchcki committed:
Fix tests after change in returned response

pawelchcki committed:
Use more correct way to setup profiling endpoint

gbbr reviewed (approved):
Thanks Pawel! LGTM!

alai97 commented on releasenotes/notes/add-trace-agent-telemetry-endpoint-f398bcd47e637081.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features:
+  - |
+    APM: Add proxy endpoint to allow Instrumentation Libraries to submit telemetry data.
```
```suggestion
    APM: Adds a proxy endpoint to allow Instrumentation Libraries to submit telemetry data.
```

alai97 reviewed (approved):
Looks good pending one copy nit!
